### DISCUSSION TITLE
"Task attached to a different loop" error in a unit test on FastApi app using SqlAlchemy AsyncSessions

### DISCUSSION THREAD

**Discussion by fyamagata** [sqlalchemy/sqlalchemy#12112] [Usage Questions]
Hi, I'm getting the following error when testing FastAPI calls, calling twice (or more) in one unittest case.  When I run sequential tests also get error.  I use async functions like the following.  It looks like a new AsyncSession is created (and destroyed) every single login api call.  SqlAlchemy documentation (https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#synopsis-core) goes `It’s advisable to invoke the AsyncEngine.dispose() method using await when using the AsyncEngine object in a scope that will go out of context and be garbage collected ...`
So, I inserted engine.dispose() at the end of get_db function, and the error wiped away.  Is this correct way to avoid the "task attached to a different loop" error?  I'm afraid of a bad side effect by disposing connection_pool this way.
Note that live server does not get this error.  Only unittest case gets this.  Why is it?

Extra notes
* Database engine is postgresql
* AsyncClient of httpx did not solve the error
* Tried several other measures, but only insertion of engine.dispose() solved the error, so far

### target code

login.py (a FastApi router)
```
from fastapi import APIRouter, Depends, HTTPException, Response, status
from app.api.deps import get_db
--8<-- snip --8<--

router = APIRouter(route_class=LoggingRoute)

@router.post("/login", response_model=str, status_code=status.HTTP_200_OK)
async def login(
    token_request: TokenRequest,
    db: Annotated[AsyncSession, Depends(get_db)],
    response: Response,
):
--8<-- snip --8<--
```

deps.py
```
from sqlalchemy.ext.asyncio import AsyncSession, async_sessionmaker, create_async_engine
--8<-- snip --8<--

engine = create_async_engine(
    settings.SQLALCHEMY_DATABASE_URI,
    pool_pre_ping=True,
)

db_session = async_sessionmaker(
    bind=engine,
    expire_on_commit=False,
    class_=AsyncSession,
)

async def get_db():
    db = db_session()
    try:
        yield db
    finally:
        await db.close()
        # ↓ this avoids "task attached to a different loop" error
        # await engine.dispose() 
```

### test code

test_login.py
```
import unittest
from fastapi.testclient import TestClient
from app.main import app
--8<-- snip --8<--

def test_login(TeseCase):
    request_json1 = {"user":"u1111" ,"password": "p1111"}
    request_json2 = {"user":"u2222" ,"password": "p2222"}
    client = TestClient(app)
    response = self.client.post("/api/login", json=request_json1)
    self.assertEqual(response.status_code, 200)
    # call the api second time -> get error
    response = self.client.post("/api/login", json=request_json2)
    self.assertEqual(response.status_code, 200)
```

error message
```
RuntimeError: Task <Task pending name='anyio.from_thread.BlockingPortal._call_func' coro=<BlockingPortal._call_func() 
running at /usr/local/lib/python3.12/site-packages/anyio/from_thread.py:221> cb=[TaskGroup._spawn.<locals>.task_done() at
/usr/local/lib/python3.12/site-packages/anyio/_backends/_asyncio.py:756]> got Future <Future pending cb=
[BaseProtocol._on_waiter_completed()]> attached to a different loop
```

**Comment by CaselIT** [comment]
Hi,

I'm assuming you are using pytest-asyncio. If I'm not mistaken it will create a new loop for each test by default. You should look into the docs so that it doesn't do that

**Reply by fyamagata** [reply]
@CaselIT 
Thank you for reply.  I'm using python unittest module (easy to learn to C user), I've also tried pytest but things didn't turn better. Do I have to use pytest-asyncio? OK, I'm looking into loop issue.  So, you mean engine.dispose will cause a result something bad?  I didn't face anything bad, though.

**Reply by CaselIT** [reply]
No, I assumed pytest since most people use it.

>  So, you mean engine.dispose will cause a result something bad? I didn't face anything bad, though.

You should not need to dispose the engine after each test. If you have to to stop this issue it seems that each test runs in its own loop.

**Reply by fyamagata** [reply]
@CaselIT 
I uses almost the same python like unittest framework on C and Objective C.  So, I prefer it.
From my understanding, engine.dispose() will dispose its connection pool, rather than engine itself.  No dispose after each api call means each user shares the same connection pool.  This feature sounds like insecure to me.  This makes me misunderstand that disposing connection pool every time meaningful.  Thank you for your help.

**Reply by fyamagata** [reply]
@CaselIT 
Puzzled, 
> it will create a new loop for each test by default

But, I get error on second api call within one test case method, not "second test".  Is it the same thing?

**Reply by CaselIT** [reply]
> From my understanding, engine.dispose() will dispose its connection pool, rather than engine itself. No dispose after each api call means each user shares the same connection pool. This feature sounds like insecure to me. This makes me misunderstand that disposing connection pool every time meaningful. Thank you for your help.

It makes no sense to dispose the pool after every api call. If you don't want to use connection pooling you can change the pool to use null pool.
not sure I follow why it should be insecure though. It's a connection to the db, usually the same db is used by all api calls. 

Not sure if you already have the issue or you solved it.

If you need further assistance I think having a runnable example would be better.

**Comment by fyamagata** [comment]
@CaselIT 
I found `IsolatedAsyncioTestCase` test case class. https://docs.python.org/3.13/library/unittest.html#unittest.IsolatedAsyncioTestCase
I suspect this test case will address loop issue, accordingly.  I will try.  Thanks a lot!!

**Reply by aliwo** [reply]
Hi! I'm having same issue! did you managed to solve this with `IsolatedAsyncioTestCase`?
